dbms/src/Encryption/KeyspacesKeyManager.cpp

```diff
@@ -0,0 +1,147 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Encryption/KeyspacesKeyManager.h>
+#include <IO/MemoryReadWriteBuffer.h>
+#include <Poco/Path.h>
+#include <Storages/KVStore/FFI/FileEncryption.h>
+#include <Storages/KVStore/FFI/ProxyFFI.h>
+#include <Storages/Page/V3/Universal/UniversalPageStorage.h>
+#include <Storages/Page/V3/Universal/UniversalWriteBatchImpl.h>
+#include <openssl/md5.h>
+
+
+namespace DB
+{
+
+KeyspacesKeyManager::KeyspacesKeyManager(EngineStoreServerWrap * tiflash_instance_wrap_)
+    : tiflash_instance_wrap(tiflash_instance_wrap_)
+    , keyspace_id_to_key(1024, 1024)
+    , master_key(std::make_unique<MasterKey>(tiflash_instance_wrap->proxy_helper->getMasterKey()))
+{}
+
+FileEncryptionInfo KeyspacesKeyManager::getInfo(const EncryptionPath & ep)
+{
+    const auto keyspace_id = ep.keyspace_id;
+    if (unlikely(keyspace_id == NullspaceID)
+        || (likely(!tiflash_instance_wrap->proxy_helper->getKeyspaceEncryption(keyspace_id))))
+    {
+        return FileEncryptionInfo{
+            FileEncryptionRes::Disabled,
+            EncryptionMethod::Plaintext,
+            nullptr,
+            nullptr,
+            nullptr,
+        };
+    }
+
+    RUNTIME_CHECK(ps_write != nullptr);
+    auto load_func = [this, keyspace_id]() -> EncryptionKeyPtr {
+        const auto page_id = UniversalPageIdFormat::toFullPageId(
+            UniversalPageIdFormat::toLocalKVPrefix(UniversalPageIdFormat::LocalKVKeyType::EncryptionKey, keyspace_id),
+            ENCRYPTION_KEY_RESERVED_PAGEU64_ID);
```

Comment:
Better define this key format in `UniversalPageIdFormatImpl.h`

Reply:
And I think we don't need to append a `ENCRYPTION_KEY_RESERVED_PAGEU64_ID` to its end

Reply:
done